crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -34,27 +43,33 @@ impl WasiRuntime {
     /// * `dirs` - a map of local file system paths to optional path names in the runtime
     ///     (e.g. /tmp/foo/myfile -> /app/config). If the optional value is not given,
     ///     the same path will be allowed in the runtime
-    /// * `log_file_location` - location for storing logs
-    pub fn new<M: AsRef<Path>>(
+    /// * `log_dir` - location for storing logs
+    pub fn new<M: AsRef<Path>, L: AsRef<Path>>(
         module_path: M,
         env: HashMap<String, String>,
         args: Vec<String>,
         dirs: HashMap<String, Option<String>>,
+        log_dir: L,
     ) -> Result<Self, failure::Error> {
         let module_data = wat::parse_file(module_path)?;
 
+        // We need to use named temp file because we need multiple file handles
+        // and if we are running in the temp dir, we run the possibility of the
+        // temp file getting cleaned out from underneath us while running. If we
+        // think it necessary, we can make these permanent files with a cleanup
+        // loop that runs elsewhere. These will get deleted when the reference
+        // is dropped
         Ok(WasiRuntime {
-            module_data,
+            module_data: Arc::new(module_data),
             env,
             args,
             dirs,
+            output: NamedTempFile::new_in(log_dir)?,
         })
     }
 
-    pub fn run(&self, output: File) -> Result<(), failure::Error> {
-        let engine = wasmtime::Engine::default();
-        let store = wasmtime::Store::new(&engine);
-
+    pub async fn start(&self) -> Result<RuntimeHandle<tokio::fs::File>, failure::Error> {
+        let output = self.output.reopen()?;
```

Comment:
Do these calls to `reopen` also block? We might also need to wrap these in `spawn_blocking`